File: tools/encoding/emit_candidate_selection_scb0.c
/* Experimental SCB-0 candidate-selection fixture emitter. Not a compiler. */
#include <stdint.h>
#include <stdio.h>
#include <string.h>

struct buffer {
    unsigned char bytes[4096];
    size_t length;
    int failed;
};

typedef void (*type_emitter)(struct buffer *);

static void
put_raw(struct buffer *buffer, const unsigned char *bytes, size_t length)
{
    if (length > sizeof buffer->bytes - buffer->length) {
        buffer->failed = 1;
        return;
    }
    memcpy(buffer->bytes + buffer->length, bytes, length);
    buffer->length += length;
}

static void
put_u8(struct buffer *buffer, uint8_t value)
{
    put_raw(buffer, &value, 1);
}

static void
put_u32(struct buffer *buffer, uint32_t value)
{
    const unsigned char bytes[4] = {
        (unsigned char)(value >> 24), (unsigned char)(value >> 16),
        (unsigned char)(value >> 8), (unsigned char)value
    };
    put_raw(buffer, bytes, sizeof bytes);
}

static void
put_name(struct buffer *buffer, const char *value)
{
    const size_t length = strlen(value);
    put_u32(buffer, (uint32_t)length);
    put_raw(buffer, (const unsigned char *)value, length);
}

static void
put_local_type_ref(struct buffer *buffer, uint32_t index)
{
    put_u8(buffer, UINT8_C(0));
    put_u32(buffer, index);
}

static void
put_variant_ref(struct buffer *buffer, uint32_t tag)
{
    put_local_type_ref(buffer, UINT32_C(4));
    put_u32(buffer, tag);
}

static void
put_field_ref(struct buffer *buffer, uint32_t owner, uint32_t index)
{
    put_u8(buffer, UINT8_C(0));
    put_local_type_ref(buffer, owner);
    put_u32(buffer, index);
}

static void type_unit(struct buffer *b) { put_u8(b, UINT8_C(0)); }
static void type_bool(struct buffer *b) { put_u8(b, UINT8_C(1)); }
static void type_u64(struct buffer *b) { put_u8(b, UINT8_C(5)); }

static void
put_declared_type(struct buffer *buffer, uint32_t index)
{
    put_u8(buffer, UINT8_C(21));
    put_local_type_ref(buffer, index);
}

static void type_candidate(struct buffer *b) { put_declared_type(b, 0); }
static void type_candidate_id(struct buffer *b) { put_declared_type(b, 1); }
static void type_epoch(struct buffer *b) { put_declared_type(b, 2); }
static void type_input(struct buffer *b) { put_declared_type(b, 3); }
static void type_rejection(struct buffer *b) { put_declared_type(b, 4); }

static void
type_candidate_array(struct buffer *buffer)
{
    put_u8(buffer, UINT8_C(18));
    type_candidate(buffer);
    put_u32(buffer, UINT32_C(32));
}

static void
type_option_candidate(struct buffer *buffer)
{
    put_u8(buffer, UINT8_C(15));
    type_candidate(buffer);
}

static void
type_selection(struct buffer *buffer)
{
    put_u8(buffer, UINT8_C(16));
    type_option_candidate(buffer);
    type_unit(buffer);
}

static void
type_kernel_result(struct buffer *buffer)
{
    put_u8(buffer, UINT8_C(19));
    type_candidate(buffer);
    type_rejection(buffer);
}

static void
expr_local(struct buffer *buffer, type_emitter type, uint32_t index)
{
    type(buffer);
    put_u8(buffer, UINT8_C(4));
    put_u32(buffer, index);
}

static void
expr_project_local(struct buffer *buffer, type_emitter result_type,
    type_emitter record_type, uint32_t local_index, uint32_t owner,
    uint32_t field_index)
{
    result_type(buffer);
    put_u8(buffer, UINT8_C(7));
    expr_local(buffer, record_type, local_index);
    put_field_ref(buffer, owner, field_index);
}

static void
expr_equal_projects(struct buffer *buffer,
    type_emitter left_type, type_emitter left_record_type,
    uint32_t left_local, uint32_t left_owner, uint32_t left_field,
    type_emitter right_type, type_emitter right_record_type,
    uint32_t right_local, uint32_t right_owner, uint32_t right_field)
{
    type_bool(buffer);
    put_u8(buffer, UINT8_C(14));
    expr_project_local(buffer, left_type, left_record_type, left_local,
        left_owner, left_field);
    expr_project_local(buffer, right_type, right_record_type, right_local,
        right_owner, right_field);
}

static void
expr_rejection(struct buffer *buffer, uint32_t tag)
{
    type_rejection(buffer);
    put_u8(buffer, UINT8_C(8));
    put_variant_ref(buffer, tag);
    put_u32(buffer, UINT32_C(0));
}

static void
expr_predicate(struct buffer *buffer)
{
    expr_equal_projects(buffer,
        type_candidate_id, type_candidate, 0, 0, 2,
        type_candidate_id, type_input, 1, 3, 2);
}

static void
expr_selection(struct buffer *buffer)
{
    type_selection(buffer);
    put_u8(buffer, UINT8_C(29));
    expr_project_local(buffer, type_candidate_array, type_input, 0, 3, 0);
    put_u32(buffer, UINT32_C(1));
    type_candidate(buffer);
    type_bool(buffer);
    expr_predicate(buffer);
}

static void
constructor_result(struct buffer *buffer, uint32_t tag)
{
    put_u8(buffer, UINT8_C(2));
    type_option_candidate(buffer);
    type_unit(buffer);
    put_u32(buffer, tag);
}

static void
constructor_option(struct buffer *buffer, uint32_t tag)
{
    put_u8(buffer, UINT8_C(1));
    type_candidate(buffer);
    put_u32(buffer, tag);
}

static void
kernel_reject(struct buffer *buffer, uint32_t tag, uint32_t precedence)
{
    put_u8(buffer, UINT8_C(1));
    expr_rejection(buffer, tag);
    put_u32(buffer, precedence);
}

static void
kernel_accept(struct buffer *buffer)
{
    put_u8(buffer, UINT8_C(0));
    expr_local(buffer, type_candidate, 0);
}

static void
kernel_some_arm(struct buffer *buffer)
{
    put_u8(buffer, UINT8_C(2));
    expr_equal_projects(buffer,
        type_epoch, type_candidate, 0, 0, 1,
        type_epoch, type_input, 3, 3, 1);
    expr_rejection(buffer, 3);
    put_u32(buffer, UINT32_C(2));

    put_u8(buffer, UINT8_C(2));
    expr_project_local(buffer, type_bool, type_candidate, 0, 0, 0);
    expr_rejection(buffer, 4);
    put_u32(buffer, UINT32_C(3));
    kernel_accept(buffer);
}

static void
kernel_option_match(struct buffer *buffer)
{
    put_u8(buffer, UINT8_C(5));
    expr_local(buffer, type_option_candidate, 0);
    put_u32(buffer, UINT32_C(2));
    constructor_option(buffer, 0);
    kernel_reject(buffer, 1, 0);
    constructor_option(buffer, 1);
    kernel_some_arm(buffer);
}

static void
kernel_result_match(struct buffer *buffer)
{
    put_u8(buffer, UINT8_C(5));
    expr_local(buffer, type_selection, 0);
    put_u32(buffer, UINT32_C(2));
    constructor_result(buffer, 0);
    kernel_option_match(buffer);
    constructor_result(buffer, 1);
    kernel_reject(buffer, 2, 1);
}

static void
kernel_expression(struct buffer *buffer)
{
    put_u8(buffer, UINT8_C(3));
    expr_selection(buffer);
    kernel_result_match(buffer);
}

static void
put_resource_bounds(struct buffer *buffer, uint32_t steps, uint32_t live,
    uint32_t depth, uint32_t workspace)
{
    put_u32(buffer, steps);
    put_u32(buffer, live);
    put_u32(buffer, depth);
    put_u32(buffer, workspace);
}

static void
put_record_field(struct buffer *buffer, const char *field_name,
    type_emitter field_type)
{
    put_name(buffer, field_name);
    field_type(buffer);
}

static void
put_type_declarations(struct buffer *buffer)
{
    put_u32(buffer, UINT32_C(5));

    put_name(buffer, "Candidate");
    put_u8(buffer, UINT8_C(2));
    put_u32(buffer, UINT32_C(3));
    put_record_field(buffer, "enabled", type_bool);
    put_record_field(buffer, "epoch", type_epoch);
    put_record_field(buffer, "id", type_candidate_id);

    put_name(buffer, "CandidateId");
    put_u8(buffer, UINT8_C(1));
    put_u8(buffer, UINT8_C(12));
    put_u8(buffer, UINT8_C(0));
    put_u32(buffer, UINT32_C(0));
    put_u32(buffer, UINT32_C(16));

    put_name(buffer, "Epoch");
    put_u8(buffer, UINT8_C(1));
    type_u64(buffer);

    put_name(buffer, "Input");
    put_u8(buffer, UINT8_C(2));
    put_u32(buffer, UINT32_C(3));
    put_record_field(buffer, "candidates", type_candidate_array);
    put_record_field(buffer, "currentEpoch", type_epoch);
    put_record_field(buffer, "wanted", type_candidate_id);

    put_name(buffer, "Rejection");
    put_u8(buffer, UINT8_C(3));
    put_u32(buffer, UINT32_C(4));
    put_u32(buffer, UINT32_C(1)); put_name(buffer, "Missing"); put_u8(buffer, 0);
    put_u32(buffer, UINT32_C(2)); put_name(buffer, "Duplicate"); put_u8(buffer, 0);
    put_u32(buffer, UINT32_C(3)); put_name(buffer, "Stale"); put_u8(buffer, 0);
    put_u32(buffer, UINT32_C(4)); put_name(buffer, "Disabled"); put_u8(buffer, 0);
}

static void
put_kernel(struct buffer *buffer)
{
    put_name(buffer, "select");
    put_u32(buffer, UINT32_C(1));
    put_name(buffer, "input");

    put_u32(buffer, UINT32_C(1));
    type_input(buffer);
    type_kernel_result(buffer);
    put_u32(buffer, UINT32_C(4));
    put_variant_ref(buffer, 1); put_variant_ref(buffer, 2);
    put_variant_ref(buffer, 3); put_variant_ref(buffer, 4);
    kernel_expression(buffer);
    put_resource_bounds(buffer, 2048, 32768, 64, 4096);
    put_resource_bounds(buffer, 212, 19361, 8, 201);
    put_u8(buffer, UINT8_C(0));
}

static void
put_module_bounds(struct buffer *buffer)
{
    put_u32(buffer, UINT32_C(1048576));
    put_u32(buffer, UINT32_C(32));
    put_u32(buffer, UINT32_C(4096));
    put_u32(buffer, UINT32_C(65536));
    put_u32(buffer, UINT32_C(256));
    put_u32(buffer, UINT32_C(32));
    put_resource_bounds(buffer, 16777216, 8388608, 256, 8388608);
}

static struct buffer
module_payload(void)
{
    struct buffer buffer = {{0}, 0, 0};
    uint32_t index;

    put_u32(&buffer, UINT32_C(0));
    put_u32(&buffer, UINT32_C(3));
    put_name(&buffer, "seki");
    put_name(&buffer, "examples");
    put_name(&buffer, "candidate_selection");
    put_u32(&buffer, UINT32_C(1));
    put_name(&buffer, "c11_bounded");
    put_u32(&buffer, UINT32_C(1));

    put_u32(&buffer, UINT32_C(0));
    put_u32(&buffer, UINT32_C(1));
    put_name(&buffer, "CandidateIdentity");
    put_type_declarations(&buffer);
    put_u32(&buffer, UINT32_C(0));
    put_u32(&buffer, UINT32_C(1));
    put_kernel(&buffer);

    put_u32(&buffer, UINT32_C(1)); put_u32(&buffer, UINT32_C(0));
    put_u32(&buffer, UINT32_C(5));
    for (index = 0; index < 5; ++index) put_u32(&buffer, index);
    put_u32(&buffer, UINT32_C(0));
    put_u32(&buffer, UINT32_C(1)); put_u32(&buffer, UINT32_C(0));
    put_u32(&buffer, UINT32_C(5));
    for (index = 0; index < 5; ++index) put_u8(&buffer, (uint8_t)index);
    put_u32(&buffer, UINT32_C(1)); put_u8(&buffer, UINT8_C(0));
    put_module_bounds(&buffer);

    put_u32(&buffer, UINT32_C(0));
    put_u32(&buffer, UINT32_C(5));
    put_u32(&buffer, UINT32_C(1)); put_u32(&buffer, UINT32_C(2));
    put_u32(&buffer, UINT32_C(0)); put_u32(&buffer, UINT32_C(3));
    put_u32(&buffer, UINT32_C(4));
    put_u32(&buffer, UINT32_C(0));
    return buffer;
}

int
main(void)
{
    static const unsigned char magic[] = {'S', 'E', 'K', 'I'};
    const struct buffer payload = module_payload();
    struct buffer module = {{0}, 0, 0};

    if (payload.failed) return 1;
    put_raw(&module, magic, sizeof magic);
    put_u32(&module, UINT32_C(0));
    put_u8(&module, UINT8_C(0));
    put_u32(&module, (uint32_t)payload.length);
    put_raw(&module, payload.bytes, payload.length);
    if (module.failed) return 1;
    if (fwrite(module.bytes, 1, module.length, stdout) != module.length) return 1;
    return 0;
}

File: tools/encoding/emit_candidate_selection_scb0.mjs
// Experimental SCB-0 candidate-selection fixture emitter. Not a compiler.
import { createHash } from "node:crypto";

let chunks = [];
const u8 = (value) => chunks.push(Buffer.from([value]));
const u32 = (value) => {
  const bytes = Buffer.alloc(4);
  bytes.writeUInt32BE(value);
  chunks.push(bytes);
};
const raw = (value) => chunks.push(Buffer.from(value));
const name = (value) => {
  const bytes = Buffer.from(value, "ascii");
  u32(bytes.length);
  raw(bytes);
};
const sequence = (items, emit = (item) => item()) => {
  u32(items.length);
  for (const item of items) emit(item);
};
const none = () => u8(0);

const typeRef = (index) => { u8(0); u32(index); };
const domainRef = (index) => { u8(0); u32(index); };
const variantRef = (owner, tag) => { typeRef(owner); u32(tag); };
const fieldRef = (owner, index) => { u8(0); typeRef(owner); u32(index); };

const typeLeaf = (tag) => () => u8(tag);
const unitType = typeLeaf(0);
const boolType = typeLeaf(1);
const u64Type = typeLeaf(5);
const declaredType = (index) => () => { u8(21); typeRef(index); };
const identityType = (domain, length) => () => {
  u8(12); domainRef(domain); u32(length);
};
const optionType = (item) => () => { u8(15); item(); };
const resultType = (ok, error) => () => { u8(16); ok(); error(); };
const arrayType = (item, length) => () => {
  u8(18); item(); u32(length);
};
const decisionType = (accepted, rejection) => () => {
  u8(19); accepted(); rejection();
};

const candidateType = declaredType(0);
const candidateIdType = declaredType(1);
const epochType = declaredType(2);
const inputType = declaredType(3);
const rejectionType = declaredType(4);
const candidateArrayType = arrayType(candidateType, 32);
const optionCandidateType = optionType(candidateType);
const selectionType = resultType(optionCandidateType, unitType);
const kernelResultType = decisionType(candidateType, rejectionType);

const expr = (type, term) => { type(); term(); };
const local = (type, index) => () => expr(type, () => { u8(4); u32(index); });
const project = (type, record, owner, index) => () => expr(type, () => {
  u8(7); record(); fieldRef(owner, index);
});
const equal = (left, right) => () => expr(boolType, () => {
  u8(14); left(); right();
});
const variant = (tag) => () => expr(rejectionType, () => {
  u8(8); variantRef(4, tag); sequence([]);
});

const predicateBody = equal(
  project(candidateIdType, local(candidateType, 0), 0, 2),
  project(candidateIdType, local(inputType, 1), 3, 2)
);

const selectionExpr = () => expr(selectionType, () => {
  u8(29);
  project(candidateArrayType, local(inputType, 0), 3, 0)();
  sequence([candidateType], (type) => type());
  boolType();
  predicateBody();
});

const constructorResult = (tag) => {
  u8(2); optionCandidateType(); unitType(); u32(tag);
};
const constructorOption = (tag) => {
  u8(1); candidateType(); u32(tag);
};

const kernelReject = (tag, precedence) => {
  u8(1); variant(tag)(); u32(precedence);
};
const kernelAcceptCandidate = () => {
  u8(0); local(candidateType, 0)();
};

const someArm = () => {
  u8(2); // KernelRequire: stale
  equal(
    project(epochType, local(candidateType, 0), 0, 1),
    project(epochType, local(inputType, 3), 3, 1)
  )();
  variant(3)();
  u32(2);
  u8(2); // KernelRequire: disabled
  project(boolType, local(candidateType, 0), 0, 0)();
  variant(4)();
  u32(3);
  kernelAcceptCandidate();
};

const optionMatch = () => {
  u8(5);
  local(optionCandidateType, 0)();
  u32(2);
  constructorOption(0); kernelReject(1, 0);
  constructorOption(1); someArm();
};

const resultMatch = () => {
  u8(5);
  local(selectionType, 0)();
  u32(2);
  constructorResult(0); optionMatch();
  constructorResult(1); kernelReject(2, 1);
};

const kernelExpression = () => {
  u8(3); // KernelLet
  selectionExpr();
  resultMatch();
};

const field = (fieldName, fieldType) => {
  name(fieldName); fieldType();
};
const recordBody = (fields) => {
  u8(2); sequence(fields, ([fieldName, fieldType]) => field(fieldName, fieldType));
};
const nominalBody = (representation) => { u8(1); representation(); };
const variantCase = (tag, caseName) => {
  u32(tag); name(caseName); none();
};
const variantBody = (cases) => {
  u8(3); sequence(cases, ([tag, caseName]) => variantCase(tag, caseName));
};

const resourceBounds = (steps, live, depth, workspace) => {
  u32(steps); u32(live); u32(depth); u32(workspace);
};

const kernelBody = () => {
  sequence([inputType], (type) => type());
  kernelResultType();
  sequence([1, 2, 3, 4], (tag) => variantRef(4, tag));
  kernelExpression();
  resourceBounds(2048, 32768, 64, 4096);
  resourceBounds(212, 19361, 8, 201);
  u8(0);
};

const functionKey = (base, labels) => {
  name(base); sequence(labels, (label) => name(label));
};

const moduleBounds = () => {
  u32(1048576);  // maximum_typed_core_bytes
  u32(32);       // maximum_imports
  u32(4096);     // maximum_declarations
  u32(65536);    // maximum_expression_nodes
  u32(256);      // maximum_nesting
  u32(32);       // maximum_call_depth
  resourceBounds(16777216, 8388608, 256, 8388608);
};

const modulePayload = () => {
  chunks = [];
  u32(0); // typed-core schema version; LanguageId occupies zero octets

  sequence(["seki", "examples", "candidate_selection"], (part) => name(part));
  u32(1);
  name("c11_bounded"); u32(1);

  sequence([]); // imports
  sequence([["CandidateIdentity"]], ([key]) => { name(key); });
  sequence([
    ["Candidate", () => recordBody([
      ["enabled", boolType], ["epoch", epochType], ["id", candidateIdType]
    ])],
    ["CandidateId", () => nominalBody(identityType(0, 16))],
    ["Epoch", () => nominalBody(u64Type)],
    ["Input", () => recordBody([
      ["candidates", candidateArrayType], ["currentEpoch", epochType],
      ["wanted", candidateIdType]
    ])],
    ["Rejection", () => variantBody([
      [1, "Missing"], [2, "Duplicate"], [3, "Stale"], [4, "Disabled"]
    ])]
  ], ([key, body]) => { name(key); body(); });
  sequence([]); // functions
  sequence([["select", ["input"]]], ([base, labels]) => {
    functionKey(base, labels); kernelBody();
  });

  sequence([0], u32);             // exported domains
  sequence([0, 1, 2, 3, 4], u32);// exported types
  sequence([]);                   // exported functions
  sequence([0], u32);             // exported kernels
  sequence([0, 1, 2, 3, 4], u8); // theorem requirements
  sequence([0], u8);              // claim ceiling
  moduleBounds();
  u32(0);                         // derivation schema version
  sequence([1, 2, 0, 3, 4], u32);// type dependency order
  sequence([]);                   // function dependency order
  return Buffer.concat(chunks);
};

const payload = modulePayload();
chunks = [];
raw("SEKI"); u32(0); u8(0); u32(payload.length); raw(payload);
const moduleBytes = Buffer.concat(chunks);

if (process.argv[2] === "--hex") {
  process.stdout.write(`${moduleBytes.toString("hex")}\n`);
} else if (process.argv[2] === "--digest") {
  const hash = createHash("sha256");
  hash.update("io.frogfish.seki/module/scb0/sha256", "ascii");
  hash.update(Buffer.from([0]));
  hash.update(moduleBytes);
  process.stdout.write(`${hash.digest("hex")}\n`);
} else if (process.argv[2] === "--length") {
  process.stdout.write(`${moduleBytes.length}\n`);
} else {
  process.stdout.write(moduleBytes);
}
